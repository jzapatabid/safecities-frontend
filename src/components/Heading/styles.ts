import styled, { css } from 'styled-components'

import { HeadingProps } from '.'

export const Wrapper = styled.h1.attrs<HeadingProps>(({ level }) => ({
  as: `h${level}`
}))<HeadingProps>`
  ${({ theme, color, size, fontWeight, lineHeight }) => css`
    font-size: ${theme.font.heading[size!]};
    color: ${color};
    font-weight: ${fontWeight};
    line-height: ${lineHeight};
    @media (max-width: 550px) {
      font-size: calc(${theme.font.heading[size!]} / 1.3);
      line-height: 1.3;
    }
  `}
`
