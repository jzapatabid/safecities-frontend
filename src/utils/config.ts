// Current config -> 1 hour
export const SESSION_EXPIRY = 3
