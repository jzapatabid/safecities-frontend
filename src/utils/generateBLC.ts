/* eslint-disable @typescript-eslint/ban-ts-comment */
// @ts-ignore
import { MONTH_MAPPING } from 'constants/Global'

import * as d3 from 'd3'
import _ from 'lodash'
import {
  formatProblemPerformanceMapData,
  getLast12MonthsIncreasingOrder,
  getLast12MonthsIncreasingOrderFromGivenStartDate
} from 'utils'

const generateBLC = (value: any, serverData: any, updatedAt: string) => {
  if (!serverData || serverData?.length === 0) return
  let slope: string
  const performanceTransformed =
    typeof value === 'number'
      ? Math.sign(value) === -1
        ? `${value}`
        : Math.sign(value) === 1
        ? `+${value}`
        : 0
      : '-'

  if (performanceTransformed === 0 || performanceTransformed) {
    if (performanceTransformed === '-') {
      slope = 'unavailable'
    } else {
      slope =
        typeof performanceTransformed === 'number'
          ? 'neutral'
          : performanceTransformed.slice(0, 1) === '-'
          ? 'positive'
          : 'negative'
    }
  }

  let initialMapData
  if (updatedAt) {
    initialMapData = getLast12MonthsIncreasingOrderFromGivenStartDate(updatedAt)
  } else {
    initialMapData = getLast12MonthsIncreasingOrder()
  }

  const finalData = formatProblemPerformanceMapData(initialMapData, serverData)

  const lineData = _.cloneDeep(finalData)

  const margin = { top: 60, right: 0, bottom: 50, left: 84 }
  const width = 780 - margin.left - margin.right
  const height = 600 - margin.top - margin.bottom

  const container = document.createElement('div')

  const parseDate = (d: any) => {
    return new Date(d.year, d.month - 1)
  }

  lineData.forEach((d) => {
    d.date = parseDate(d)
    d.stateRate = d.stateRate !== undefined ? +d.stateRate : undefined
  })

  const definedBarData = finalData.filter((d: any) => d.cityRate !== 0)

  const allBarDataZero = definedBarData.length === 0

  const definedLineData = finalData.filter(
    (d: any) => d.stateRate !== undefined
  )

  const allLineDataZero =
    definedLineData.filter((item: any) => item.stateRate === 0).length ===
    definedLineData.length

  const x = d3
    .scaleBand()
    .domain(finalData.map((d) => `${MONTH_MAPPING[d.month]}, ${d.year}`))
    .range([0, width])
    .paddingInner(0.4)
    .paddingOuter(0.25)

  const y = d3
    .scaleLinear()
    .domain([
      0,
      d3.max(finalData, (d) => {
        return Math.max(d.cityRate || 0, d.stateRate || 0)
      })
    ])
    .nice()
    .range([height, 0])

  const xAxis = (g: any) =>
    g
      .attr('transform', `translate(0, ${height})`)
      .attr('class', 'x-axis')
      .call(d3.axisBottom(x).tickSize(0))
      .call((g: any) => g.select('.domain').remove())

  const yAxis = (g: any) =>
    g
      .attr('class', `y-axis`)
      .call(d3.axisLeft(y).tickSize(0))
      .call((g: any) => g.select('.domain').remove())

  d3.select(container).style('background-color', '#00000033')

  const svg = d3
    .select(container)
    .append('svg')
    .attr('width', width + margin.left + margin.right)
    .attr('height', height + margin.top + margin.bottom)
    .append('g')
    .attr('transform', `translate(${margin.left},${margin.top})`)

  svg
    .append('g')
    .selectAll('rect')
    .data(finalData)
    .enter()
    .append('rect')
    .attr('class', 'rect-box')
    .attr('x', (d) => x(`${MONTH_MAPPING[d.month]}, ${d.year}`))
    .attr('y', (d) => y(d.cityRate) || 0)
    .attr('height', (d) => height - y(d.cityRate) || 0)
    .attr('width', x.bandwidth())
    .attr('fill', (d) => (d.cityRate === 0 ? 'transparent' : '#000000'))

  svg
    .selectAll('textB')
    .data(finalData)
    .enter()
    .append('text')
    .attr('class', 'label-rect')
    .attr(
      'x',
      (d) => x(`${MONTH_MAPPING[d.month]}, ${d.year}`) + x.bandwidth() / 2
    )
    .attr('y', (d) => y(d.cityRate) - 5 || 0)
    .text((d) => {
      return d?.cityRate || ''
    })
    .style('font-size', '10px')
    .attr('fill', 'black')
    .attr('text-anchor', 'middle')

  const line = d3
    .line()
    .defined((d) => d.stateRate !== undefined)
    .x((d) => {
      return x(`${MONTH_MAPPING[d.month]}, ${d.year}`) + x.bandwidth() / 2
    })
    .y((d) => (allLineDataZero ? height : y(d.stateRate)))

  svg
    .append('path')
    .datum(definedLineData)
    .attr('class', 'line')
    .attr('d', line)
    .attr('fill', 'none')
    .attr('stroke', '#50C878')
    .attr('stroke-width', 2)

  svg
    .selectAll('.dot')
    .data(finalData)
    .enter()
    .append('circle')
    .attr('class', 'dot')
    .attr(
      'cx',
      (d) => x(`${MONTH_MAPPING[d.month]}, ${d.year}`) + x.bandwidth() / 2
    )
    .attr('cy', (d) => (allLineDataZero ? height : y(d.stateRate)))
    .attr('r', (d: any) => {
      return typeof d.stateRate === 'undefined' ? 0 : 5
    })
    .attr('fill', '#50C878')

  svg
    .selectAll('.circle-label')
    .data(finalData)
    .enter()
    .append('text')
    .attr('class', 'dot-label')
    .style('font-size', '10px')
    .attr(
      'x',
      (d) => x(`${MONTH_MAPPING[d.month]}, ${d.year}`) + x.bandwidth() / 2 - 2.5
    )
    .attr('y', (d) => (allLineDataZero ? height : y(d.stateRate) - 12.5 || 0))
    .text((d) => d.stateRate || 0)
    .attr('fill', (d) =>
      d.stateRate !== undefined && d.stateRate !== null ? '#50C878' : '#FFFFFF'
    )

  svg
    .append('g')
    .call(xAxis)
    .selectAll('.x-axis .tick > text')
    .attr('dy', '20px')
    .html('')
    .append('tspan')
    .attr('class', 'x_axis_label')
    .style('fill', 'black')
    .text((d: any) => {
      const [month] = d.split(', ')
      return `${month}`
    })

  svg
    .selectAll('.x-axis .tick > text')
    .append('tspan')
    .attr('class', 'x_axis_label')
    .attr('x', 0)
    .attr('dy', '1.2em')
    .attr('fill', 'black')
    .text((d: any) => {
      const [, year] = d.split(', ')
      return `${year}`
    })

  svg
    .append('text')
    .attr('class', 'y_axis_label')
    .style('transform', 'rotate(-90deg)')
    .attr('fill', 'black')
    .style('font-family', 'Poppins')
    .style('font-size', '12px')
    .attr('y', 0 - margin.left + 12)
    .attr('x', 0 - height / 2)
    .attr('dy', '1em')
    .text('Número de ocorrências')

  svg
    .append('g')
    .call(yAxis)
    .selectAll('.y-axis .tick > text')
    .attr(
      'transform',
      allLineDataZero && allBarDataZero ? `translate(0, ${height / 2})` : ''
    )

  // Extract the HTML string of the container
  const htmlString = container.innerHTML

  // Remove the container from the DOM
  container.remove()

  return {
    html: htmlString,
    performance: value,
    slope: slope
  }
}

export default generateBLC
