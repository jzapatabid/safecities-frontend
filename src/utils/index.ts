function getSessionExpirationHashedEpochTime(hours: number) {
  const currentTime = Date.now()
  const hoursInMilliseconds = hours * 60 * 60 * 1000
  const futureTime = currentTime + hoursInMilliseconds
  return {
    sessionExpirationEpochTime: (futureTime + 1).toString(36)
  }
}

function runAtEpochTime(epochTime: number, callback: () => void) {
  const currentTime = Date.now()
  const timeRemaining = epochTime - currentTime
  if (timeRemaining <= 0) {
    callback()
  } else {
    setTimeout(callback, timeRemaining)
  }
}

function unhashEpoch(epoch: string) {
  return parseInt(epoch, 36)
}

function getPaginationPageGroupInfo(currentPage: number) {
  const remainder = currentPage % 5
  let start, end
  if (remainder !== 0) {
    start = currentPage - (remainder - 1)
    end = currentPage + (5 - remainder)
  } else {
    end = currentPage
    start = currentPage - 4
  }
  return { start, end }
}

function getLast12MonthsIncreasingOrder() {
  const currentDate = new Date()
  const currentYear = currentDate.getFullYear()
  const currentMonth = currentDate.getMonth() + 1

  const months = []
  for (let i = 0; i < 12; i++) {
    const year = currentMonth - i <= 0 ? currentYear - 1 : currentYear
    const month = (currentMonth - i + 12) % 12 || 12

    months.push({
      year: year.toString(),
      month,
      totalCityIncidents: 0
    })
  }

  return months.reverse() // Reverse the array to get increasing order
}

function getLast12MonthsIncreasingOrderFromGivenStartDate(date: string) {
  const currentDate = new Date(date)
  const currentYear = currentDate.getFullYear()
  const currentMonth = currentDate.getMonth() + 1

  const months = []
  for (let i = 0; i < 12; i++) {
    const year = currentMonth - i <= 0 ? currentYear - 1 : currentYear
    const month = (currentMonth - i + 12) % 12 || 12

    months.push({
      year: year.toString(),
      month,
      totalCityIncidents: 0
    })
  }

  return months.reverse() // Reverse the array to get increasing order
}

function getLast5YearsQuarters() {
  const currentYear = new Date().getFullYear()
  const quarters = ['q1', 'q2', 'q3', 'q4']

  const result = []
  for (let i = 4; i >= 0; i--) {
    const year = currentYear - i
    for (const quarter of quarters) {
      result.push({
        year: year.toString(),
        quarter,
        totalCityIncidents: 0
      })
    }
  }

  return result
}

function getLast5YearsQuartersFromGivenStartDate(date: string) {
  const currentYear = new Date(date).getFullYear()
  const quarters = ['q1', 'q2', 'q3', 'q4']

  const result = []
  for (let i = 4; i >= 0; i--) {
    const year = currentYear - i
    for (const quarter of quarters) {
      result.push({
        year: year.toString(),
        quarter,
        totalCityIncidents: 0
      })
    }
  }

  return result
}

const formatProblemPerformanceMapData = (
  fillerArray: any[],
  serverDataArray: any[] = []
) => {
  const serverDataMap = new Map()
  serverDataArray.map((item) => {
    const key = item.year + '-' + item.month
    serverDataMap.set(key, item)
  })

  fillerArray.forEach((item) => {
    const key = item.year + '-' + item.month
    if (serverDataMap.has(key)) {
      const serverItem = serverDataMap.get(key)
      item.totalStateIncidents = serverItem.totalStateIncidents
      item.totalCityIncidents = serverItem.totalCityIncidents
    }
  })

  return fillerArray
}

const formatProblemTrendMapData = (
  fillerArray: any[],
  serverDataArray: any[]
) => {
  const serverDataMap = new Map()

  // Create a map of year and quarter values to the corresponding objects in the server data array
  serverDataArray.forEach((item) => {
    const key = item.year + '-' + item.quarter.toUpperCase()
    serverDataMap.set(key, item)
  })

  // Update the filler array with server data where year and quarter values collide
  fillerArray.forEach((item) => {
    item.quarter = item.quarter.toUpperCase()
    const key = item.year + '-' + item.quarter.toUpperCase()
    if (serverDataMap.has(key)) {
      const serverItem = serverDataMap.get(key)
      item.rateCityIncidents = serverItem.rateCityIncidents
      item.totalCityIncidents = serverItem.totalCityIncidents
    }
  })

  return fillerArray
}

function camelToSnake(camelCaseString: string) {
  return camelCaseString?.replace(
    /[A-Z]/g,
    (match) => `_${match.toLowerCase()}`
  )
}

function getLastYearRangeInPortuguese(date?: string) {
  if (date === 'NA') {
    return '-'
  }

  let lastYearStart
  const today = date ? new Date(date + 'T00:00:00') : new Date()

  // Exception case check
  if (today.getMonth() === 11) {
    lastYearStart = new Date(today)
    lastYearStart.setMonth(0)
  } else {
    lastYearStart = new Date(today)
    lastYearStart.setFullYear(today.getFullYear() - 1)
  }
  const lastYearEnd = new Date(today)

  // Define an array of month names in Portuguese
  const monthNames = [
    'Jan',
    'Fev',
    'Mar',
    'Abr',
    'Mai',
    'Jun',
    'Jul',
    'Ago',
    'Set',
    'Out',
    'Nov',
    'Dez'
  ]

  // Format the start and end months
  const startMonth =
    monthNames[
      lastYearStart.getMonth() === 0 && today.getMonth() !== 0
        ? lastYearStart.getMonth()
        : lastYearStart.getMonth() + 1
    ]
  const endMonth = monthNames[lastYearEnd.getMonth()]

  // Get the year of the start and end months
  const startYear = lastYearStart.getFullYear()
  const endYear = lastYearEnd.getFullYear()

  // Format the result
  const result = `${startMonth}/${startYear} - ${endMonth}/${endYear}`
  return result
}

function formatNumberToLatinAmerican(value: number): number | string {
  if (isNaN(value as number)) {
    return ''
  }

  const [integerPart, decimalPart] = Number(value)
    .toFixed(2)
    .toString()
    .split('.')

  const formattedIntegerPart = integerPart.replace(/\B(?=(\d{3})+(?!\d))/g, ',')

  const formattedDecimalPart = decimalPart.replace(/0+$/, '')

  const formattedNumber = formattedDecimalPart.length
    ? `${formattedIntegerPart}.${formattedDecimalPart}`
    : formattedIntegerPart

  return formattedNumber
}

export {
  getSessionExpirationHashedEpochTime,
  runAtEpochTime,
  unhashEpoch,
  getPaginationPageGroupInfo,
  getLast12MonthsIncreasingOrder,
  formatProblemPerformanceMapData,
  getLast5YearsQuarters,
  formatProblemTrendMapData,
  camelToSnake,
  getLastYearRangeInPortuguese,
  formatNumberToLatinAmerican,
  getLast5YearsQuartersFromGivenStartDate,
  getLast12MonthsIncreasingOrderFromGivenStartDate
}
