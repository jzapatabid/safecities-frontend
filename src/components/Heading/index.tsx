import * as S from './styles'
import theme from 'styles/theme'

export type HeadingProps = {
  color?: string
  size?: keyof typeof theme.font.heading
  fontWeight?: 400 | 500 | 600 | 700
  lineHeight?: number | string
  level?: 1 | 2 | 3 | 4 | 5 | 6
  children: React.ReactNode
}

const Heading = ({
  color = '#fff',
  size = 'regular',
  fontWeight = 600,
  level = 1,
  lineHeight,
  children
}: HeadingProps) => (
  <S.Wrapper
    level={level}
    color={color}
    size={size}
    fontWeight={fontWeight}
    lineHeight={lineHeight}
  >
    {children}
  </S.Wrapper>
)

export default Heading
