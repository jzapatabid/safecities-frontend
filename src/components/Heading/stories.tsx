import { ComponentMeta, Story } from '@storybook/react'

import Heading, { HeadingProps } from '.'

export default {
  title: 'Components/Heading',
  component: Heading,
  parameters: {
    backgrounds: {
      default: 'dark'
    }
  },
  argTypes: {
    children: {
      type: 'string'
    },
    size: {
      options: [
        'xxxsmall',
        'xxsmall',
        'xsmall',
        'small',
        'regular',
        'large',
        'xlarge',
        'xxlarge',
        'xxxlarge',
        'xxxxlarge'
      ],
      description: `xxxsmall: '1.8rem', xxsmall: '2rem', xsmall: '2.2rem', small: '2.0rem', regular: '3.6rem', large: '4.0rem', xlarge: '4.8rem', xxlarge: '6rem', xxxlarge: '7.2rem', xxxxlarge: '9.6rem'`,
      control: { type: 'select' },
      defaultValue: 'regular'
    },
    level: {
      options: [1, 2, 3, 4, 5, 6],
      description:
        'Cada level equivale a uma tag de titulo. ex: level 1 = h1, level 2 = h2, etc.',
      control: { type: 'select' },
      defaultValue: 1
    },
    color: {
      control: 'color',
      defaultValue: '#fff'
    },
    fontWeight: {
      name: 'Font Weight',
      options: [400, 500, 600, 700],
      control: { type: 'select' },
      defaultValue: 600
    },
    lineHeight: {
      name: 'Line Height'
    }
  }
} as ComponentMeta<typeof Heading>

export const Default: Story<HeadingProps> = (args) => <Heading {...args} />

Default.args = {
  children: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.'
}
