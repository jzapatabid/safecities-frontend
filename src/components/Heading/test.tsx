import { render, screen } from 'utils/test-utils'

import Heading from '.'

describe('<Heading />', () => {
  it('should render heading as h1 by default', () => {
    const { container } = render(<Heading>Bem vindo</Heading>)

    expect(
      screen.getByRole('heading', { name: /bem vindo/i, level: 1 })
    ).toBeInTheDocument()

    expect(
      screen.getByRole('heading', { name: /bem vindo/i, level: 1 })
    ).toHaveStyle({
      color: '#fff',
      'font-size': '3.6rem',
      'font-weight': 600
    })

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render heading as h2 when level 2 is passed', () => {
    render(<Heading level={2}>Bem vindo</Heading>)

    expect(
      screen.getByRole('heading', { name: /bem vindo/i, level: 2 })
    ).toBeInTheDocument()
  })

  it('should render heading as h3 when level 3 is passed', () => {
    render(<Heading level={3}>Bem vindo</Heading>)

    expect(
      screen.getByRole('heading', { name: /bem vindo/i, level: 3 })
    ).toBeInTheDocument()
  })

  it('should render heading as h4 when level 4 is passed', () => {
    render(<Heading level={4}>Bem vindo</Heading>)

    expect(
      screen.getByRole('heading', { name: /bem vindo/i, level: 4 })
    ).toBeInTheDocument()
  })

  it('should render heading as h5 when level 5 is passed', () => {
    render(<Heading level={5}>Bem vindo</Heading>)

    expect(
      screen.getByRole('heading', { name: /bem vindo/i, level: 5 })
    ).toBeInTheDocument()
  })

  it('should render heading as h5 when level 5 is passed', () => {
    render(<Heading level={5}>Bem vindo</Heading>)

    expect(
      screen.getByRole('heading', { name: /bem vindo/i, level: 5 })
    ).toBeInTheDocument()
  })

  it('should render heading as h6 when level 6 is passed', () => {
    render(<Heading level={6}>Bem vindo</Heading>)

    expect(
      screen.getByRole('heading', { name: /bem vindo/i, level: 6 })
    ).toBeInTheDocument()
  })

  it('should render heading with the color "white" when color white is passed', () => {
    render(<Heading color="white">Bem vindo</Heading>)

    expect(
      screen.getByRole('heading', { name: /bem vindo/i, level: 1 })
    ).toHaveStyle({
      color: 'white'
    })
  })

  it('should render heading with size "4.0rem" when size large is passed', () => {
    render(<Heading size="large">Bem vindo</Heading>)

    expect(
      screen.getByRole('heading', { name: /bem vindo/i, level: 1 })
    ).toHaveStyle({
      'font-size': '4.0rem'
    })
  })

  it('should render heading with fontWeight "400" when passed', () => {
    render(<Heading fontWeight={400}>Bem vindo</Heading>)

    expect(
      screen.getByRole('heading', { name: /bem vindo/i, level: 1 })
    ).toHaveStyle({
      'font-weight': 400
    })
  })
})
