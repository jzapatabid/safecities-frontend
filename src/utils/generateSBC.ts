/* eslint-disable @typescript-eslint/ban-ts-comment */
// @ts-ignore
import * as d3 from 'd3'
import {
  formatProblemTrendMapData,
  getLast5YearsQuarters,
  getLast5YearsQuartersFromGivenStartDate
} from 'utils'

import dataVizStyles from '../components/StackedBarChart/datavizStyles.module.css'

import { getCriticalityDetail } from 'components/DetailCardV2'

const groupKey = 'year'
const stackKey = 'quarter'

const colors = {
  'Santa Catarina': '#000000',
  Florianópolis: '#000000',
  totalCityIncidents: '#000000'
}

const keys = ['totalCityIncidents']

const generateSBC = (
  value: any,
  serverData: any,
  measurement_unit: any,
  updatedAt: string
) => {
  if (!serverData || serverData?.length === 0) return
  let slope: string, criticality_info: { index: number; criticality: string }
  let data: Iterable<any>

  if (updatedAt) {
    data = formatProblemTrendMapData(
      getLast5YearsQuartersFromGivenStartDate(updatedAt),
      serverData
    )
  } else {
    data = formatProblemTrendMapData(getLast5YearsQuarters(), serverData)
  }

  // const data = formatProblemTrendMapData(getLast5YearsQuarters(), serverData)

  const trendTransformed =
    typeof value === 'number'
      ? Math.sign(value) === -1
        ? `${value}`
        : Math.sign(value) === 1
        ? `+${value}`
        : 0
      : '-'

  if (trendTransformed === 0 || trendTransformed) {
    if (trendTransformed === '-') {
      slope = 'unavailable'
    } else {
      slope =
        typeof trendTransformed === 'number'
          ? 'neutral'
          : trendTransformed.slice(0, 1) === '-'
          ? 'positive'
          : 'negative'
    }

    criticality_info = getCriticalityDetail(slope)
  }

  // Create a container element
  const container = document.createElement('div')

  d3.select(container).style('background-color', '#00000033')

  //graphic code
  const width = 780
  const height = 550

  const margin = {
    top: 30,
    right: 60,
    bottom: 60,
    left: 70
  }

  const definedBarData = data.filter((d: any) => d.totalCityIncidents !== 0)

  const allBarDataZero = definedBarData.length === 0

  const definedLineData = data.filter(
    (d: any) =>
      d.rateCityIncidents !== undefined && d.rateCityIncidents !== null
  )

  const allLineDataZero =
    definedLineData.filter((item: any) => item.rateCityIncidents === 0)
      .length === definedLineData.length

  const dataByYear = d3.group(data, (d: any) => d[groupKey])
  const dataByQuarter = d3.group(data, (d: any) => d[stackKey])

  const stackGenerator = d3.stack().keys(keys)
  const layers = stackGenerator(data)

  const svg = d3
    .select(container)
    .append('svg')
    .attr('id', 'rootDS')
    .attr('height', height)
    .attr('width', width)

  const outerGroupXScale = d3
    .scaleBand()
    .domain(dataByYear.keys())
    .range([margin.left, width - margin.right])
    .padding(0.05)
  const outerGroupXAxis = (g: any) => {
    g.attr('transform', `translate(0, ${height - margin.bottom / 2 - 16})`)
      .attr('class', 'x-axis')
      .call(d3.axisBottom(outerGroupXScale))
      .call((g: any) => g.selectAll('.domain').remove())
      .call((g: any) => {
        g.selectAll('.tick>line').remove()
      })
      .call((g: any) => {
        g.selectAll('.tick>text')
          .attr('class', dataVizStyles.tick_text)
          .style('font-family', 'Poppins')
          .style('font-size', '10px')
      })
      .call((g: any) => g.select('.domain').attr('d', 'M180,0H820'))
  }

  const innerGroupXScale = d3
    .scaleBand()
    .domain(dataByQuarter.keys())
    .range([0, outerGroupXScale.bandwidth()])
    .padding(0.25)
    .paddingInner(0.25)
    .paddingOuter(0.5)
  const innerGroupXAxis = (g: any) => {
    g.attr('transform', `translate(0, ${height - margin.bottom})`)
      .attr('class', 'x-axis')
      .call(d3.axisBottom(innerGroupXScale))
      .attr('font-size', '10px')
      .call((g: any) => g.selectAll('.domain').remove())
      .call((g: any) => {
        g.selectAll('.tick>line').remove()
      })
      .call((g: any) => {
        g.selectAll('.tick>text')
          .attr('class', dataVizStyles.tick_text)
          .style('font-family', 'Poppins')
          .style('font-size', '10px')
      })
  }

  const extent = [
    0.9 * d3.min(layers, (layer) => d3.min(layer, (d) => d[1])),
    1.1 * d3.max(layers, (layer) => d3.max(layer, (d) => d[1]))
  ]
  const [yMin] = extent
  const yScale = d3
    .scaleLinear()
    .domain(extent)
    .range([height - margin.bottom, margin.top])

  const yAxisLeft = (g: any) => {
    g.attr('transform', `translate(${margin.left}, 0)`)
      .attr('class', dataVizStyles.y_axis)
      .call(d3.axisLeft(yScale))
      .call((g: any) => g.selectAll('.domain').remove())
      .call((g: any) => {
        g.selectAll('.tick>line').remove()
      })
      .call((g: any) => {
        g.selectAll('.tick>text')
          .attr('class', dataVizStyles.tick_text)
          .style('font-family', 'Poppins')
          .style('font-size', '10px')
          .attr(
            'transform',
            allBarDataZero
              ? `translate(10, ${height / 2 - margin.bottom})`
              : 'translate(10, 0)'
          )
      })
  }

  const yAxisLabel = (g: any) => {
    g.append('text')
      .attr('transform', `rotate(-90, -45, ${height / 1.66})`)
      .attr('class', dataVizStyles.y_axis_label)
      .style('fill', '#000000')
      .style('font-family', 'Poppins')
      .style('font-size', '10px')
      .attr('x', 65)
      .attr('y', height / 1.66)
      .text('Número de ocorrências')
  }

  const yAxisLabelLinePlot = (g: any) => {
    g.append('text')
      .attr('transform', `rotate(90, 45, ${height / 2.6})`)
      .attr('class', `${dataVizStyles.y_axis_label} line_y_axis`)
      .style('fill', '#000000')
      .style('font-family', 'Poppins')
      .style('font-size', '10px')
      .attr('x', 40)
      .attr('y', height / 2.65)
      .text(measurement_unit)
  }

  const lineYScale = d3
    .scaleLinear()
    .domain([0, d3.max(data, (d) => d.rateCityIncidents)])
    .range([height - margin.bottom, margin.top])

  const lineYAxis = (g: any) => {
    g.attr('transform', `translate(${width - margin.right}, 0)`)
      .attr('class', `y_axis_line`)
      .call(d3.axisRight(lineYScale))
      .call((g: any) => g.selectAll('.domain').remove())
      .call((g: any) => {
        g.selectAll('.tick>line').remove()
      })
      .call((g: any) => {
        g.selectAll('.tick>text')
          .attr('class', dataVizStyles.tick_text)
          .style('font-family', 'Poppins')
          .style('font-size', '10px')
          .attr(
            'transform',
            allLineDataZero ? `translate(0, ${height / 2 - margin.bottom})` : ''
          )
      })
  }

  const combinedData = data.map((d: any) => ({
    [groupKey]: d[groupKey],
    [stackKey]: d[stackKey],
    values: [
      { key: keys[0], value: d[keys[0]] },
      { key: keys[1], value: d[keys[1]] }
    ],
    rateCityIncidents: d.rateCityIncidents
  }))

  // const tooltip = d3
  //   .select('#dataViz')
  //   .append('div')
  //   .attr('class', 'tooltip')
  //   .style('opacity', 0)

  const stateG = svg
    .selectAll('.state')
    .data(dataByYear)
    .join('g')
    .attr('class', 'state')
    .attr('fill', (d) => colors[d[0]])
    .attr('transform', (d) => `translate(${outerGroupXScale(d[0])}, 0)`)

  stateG.append('g').attr('class', 'stateAxis').call(innerGroupXAxis)

  const yearG = stateG
    .selectAll('.yearG')
    .data((d) => {
      const filteredByState = data.filter((i: any) => i[groupKey] === d[0])
      const groupedByQuarter = d3.group(
        filteredByState,
        (a: any) => a[stackKey]
      )
      return groupedByQuarter
    })
    .join('g')
    .attr('class', 'yearG')
    .attr('transform', (d) => {
      return `translate(${innerGroupXScale(d[0])}, 0)`
    })

  yearG
    .selectAll('.layers')
    .data((d) => {
      return stackGenerator(d[1])
    })
    .join('rect')
    .attr('class', 'layers')
    .attr('y', (d) => yScale(d[0][1]))
    .attr('fill', (d) => colors[d.key])
    .attr('width', () => innerGroupXScale.bandwidth())
    .attr('height', (d) => {
      const lower = d[0][0]
      const upper = d[0][1]
      const firstBarAdjustment = lower === 0 ? yMin : 0
      return yScale(lower + firstBarAdjustment) - yScale(upper)
    })

  yearG
    .selectAll('.textValue')
    .data((d) => {
      return stackGenerator(d[1])
    })
    .join('text')
    .attr('class', 'layers2')
    .attr('y', (d) => yScale(d[0][1]) - 5)
    .attr('fill', () => 'black')
    .text((d) => {
      return d.key === keys[0]
        ? `${d[0].data[keys[0]] !== 0 ? d[0].data[keys[0]] : ''}`
        : `${d[0].data[keys[1]]}`
    })
    .style('font-family', 'Poppins')
    .style('font-size', '10px')
    .style('text-anchor', 'right')

  const line = d3
    .line()
    .defined((d) => d.rateCityIncidents !== undefined)
    .x(
      (d) =>
        outerGroupXScale(d[groupKey]) +
        innerGroupXScale(d[stackKey]) +
        innerGroupXScale.bandwidth() / 2
    )
    .y((d) =>
      allLineDataZero ? height - margin.bottom : lineYScale(d.rateCityIncidents)
    )

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const linePath = svg
    .append('path')
    .datum(
      combinedData.filter((item: any) => item.rateCityIncidents !== undefined)
    )
    .attr('class', 'line-path')
    .attr('fill', 'none')
    .attr('stroke-width', 2)
    .attr('stroke', '#50C878')
    .attr('d', line)

  svg
    .selectAll('.circle')
    .data(combinedData)
    .enter()
    .append('circle')
    .attr('class', 'circle')
    .attr(
      'cx',
      (d) =>
        outerGroupXScale(d[groupKey]) +
        innerGroupXScale(d[stackKey]) +
        innerGroupXScale.bandwidth() / 2
    )
    .attr('cy', (d: any) =>
      allLineDataZero ? height - margin.bottom : lineYScale(d.rateCityIncidents)
    )
    .attr('r', (d) =>
      d.rateCityIncidents !== undefined && d.rateCityIncidents !== null ? 5 : 0
    )
    .attr('fill', '#50C878')

  svg
    .selectAll('.textA')
    .data(combinedData)
    .enter()
    .append('text')
    .attr('class', 'circle-label')
    .attr(
      'x',
      (d) =>
        outerGroupXScale(d[groupKey]) +
        innerGroupXScale(d[stackKey]) +
        innerGroupXScale.bandwidth() / 2 -
        3.5
    )
    .attr('y', (d: any) =>
      allLineDataZero
        ? height - margin.bottom
        : lineYScale(d.rateCityIncidents) - 12.5 || 0
    )
    .text((d) =>
      d.rateCityIncidents !== undefined && d.rateCityIncidents !== null
        ? `${`${d.rateCityIncidents}`.slice(0, 4)}`
        : 0
    )
    .attr('fill', (d) =>
      d.rateCityIncidents !== undefined && d.rateCityIncidents !== null
        ? '#50C878'
        : '#FFFFFF'
    )
    .style('text-anchor', 'middle')
    .style('font-size', '10px')

  svg.append('g').call(outerGroupXAxis)
  svg.append('g').call(yAxisLeft).call(yAxisLabel)
  svg.append('g').call(lineYAxis).call(yAxisLabelLinePlot)

  svg.node()

  // Extract the HTML string of the container
  const htmlString = container.innerHTML

  // Remove the container from the DOM
  container.remove()

  return {
    html: htmlString,
    trend: trendTransformed,
    slope: slope,
    criticality_info: criticality_info
  }
}

export default generateSBC
