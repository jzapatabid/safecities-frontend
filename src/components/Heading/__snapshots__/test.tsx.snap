// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Heading /> should render heading as h1 by default 1`] = `
.c0 {
  font-size: 3.6rem;
  color: #fff;
  font-weight: 600;
}

@media (max-width:550px) {
  .c0 {
    font-size: calc(3.6rem / 1.3);
    line-height: 1.3;
  }
}

<h1
  class="c0"
  color="#fff"
  font-weight="600"
>
  Bem vindo
</h1>
`
